src/control/server/ctl_storage_rpc.go

```diff
@@ -325,92 +327,145 @@ func (c *ControlService) StorageScan(ctx context.Context, req *ctlpb.StorageScan
-	instances := c.harness.Instances()
-	resp := new(ctlpb.StorageFormatResp)
-	resp.Mrets = make([]*ctlpb.ScmMountResult, 0, len(instances))
-	resp.Crets = make([]*ctlpb.NvmeControllerResult, 0, len(instances))
-	scmChan := make(chan *ctlpb.ScmMountResult, len(instances))
-	mdFormatted := false
-
+func (c *ControlService) formatMetadata(instances []Engine, reformat bool) (bool, error) {
 	// Format control metadata first, if needed
 	if needs, err := c.storage.ControlMetadataNeedsFormat(); err != nil {
-		return nil, errors.Wrap(err, "detecting if metadata format is needed")
-	} else if needs || req.Reformat {
+		return false, errors.Wrap(err, "detecting if metadata format is needed")
+	} else if needs || reformat {
 		engineIdxs := make([]uint, len(instances))
 		for i, eng := range instances {
 			engineIdxs[i] = uint(eng.Index())
 		}
 
 		c.log.Debug("formatting control metadata storage")
 		if err := c.storage.FormatControlMetadata(engineIdxs); err != nil {
-			return nil, errors.Wrap(err, "formatting control metadata storage")
+			return false, errors.Wrap(err, "formatting control metadata storage")
+		}
+
+		return true, nil
+	}
+
+	c.log.Debug("no control metadata format needed")
+	return false, nil
+}
+
+func checkTmpfsMem(log logging.Logger, scmCfgs map[int]*storage.TierConfig, getMemInfo func() (*common.MemInfo, error)) error {
+	if scmCfgs[0].Class != storage.ClassRam {
+		return nil
+	}
+
+	var memRamdisks uint64
+	for _, sc := range scmCfgs {
+		memRamdisks += uint64(sc.Scm.RamdiskSize) * humanize.GiByte
+	}
+
+	mi, err := getMemInfo()
+	if err != nil {
+		return errors.Wrap(err, "retrieving system meminfo")
+	}
+	memAvail := uint64(mi.MemAvailableKiB) * humanize.KiByte
+
+	if err := checkMemForRamdisk(log, memRamdisks, memAvail); err != nil {
+		return errors.Wrap(err, "check ram available for all tmpfs")
+	}
+
+	return nil
+}
+
+type formatScmReq struct {
+	log        logging.Logger
+	reformat   bool
+	instances  []Engine
+	getMemInfo func() (*common.MemInfo, error)
+}
+
+func formatScm(ctx context.Context, req formatScmReq, resp *ctlpb.StorageFormatResp) (map[int]string, map[int]bool, error) {
+	needFormat := make(map[int]bool)
+	scmCfgs := make(map[int]*storage.TierConfig)
+	allNeedFormat := true
+
+	for idx, ei := range req.instances {
+		needs, err := ei.GetStorage().ScmNeedsFormat()
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "detecting if SCM format is needed")
+		}
+		if !needs {
+			allNeedFormat = false
+		}
+		needFormat[idx] = needs
+
+		scmCfg, err := ei.GetStorage().GetScmConfig()
+		if err != nil || scmCfg == nil {
+			return nil, nil, errors.Wrap(err, "retrieving SCM config")
+		}
+		scmCfgs[idx] = scmCfg
+	}
+
+	if allNeedFormat {
+		// Check available RAM is sufficient before formatting SCM on engines.
+		if err := checkTmpfsMem(req.log, scmCfgs, req.getMemInfo); err != nil {
+			return nil, nil, err
```

Comment:
A separate check is performed on engine start https://github.com/daos-stack/daos/pull/11978/files#diff-d4337e9b4e14476c4e9211f64d51adcdfc20fe8f35ce486e450063d637ef1d7cR530 that validates there is enough memory to support the single engine's ram-disk. This check on first format validates that there will be enough for ram-disks of all engines whereas when we do the per-engine check we don't have visibility on the ram-disk(s) of any other engines.

Reply:
but yes, this could be modified to verify any of the subsets being formatted, I will keep that in mind for future refactors. In MD-on-SSD phase 2 we will probably get rid of the concept of SCM format when in MD-on-SSD mode and just perform the check on start. The only problem then is that when checking available memory we need to be looking at the capacity usage of each engine's tmpfs to make sure we have enough available memory to cover all the unused capacities.

Reply:
I see. Thanks for the thorough explanation!